test/gurka/test_build_admin.cc

```diff
@@ -0,0 +1,183 @@
+                                      {{{{gurka::way_member, "ABCDEFA", "outer"}}},
+                                       {{"type", "boundary"},
+                                        {"boundary", "administrative"},
+                                        {"admin_level", "2"},
+                                        {"name", "USA"}}}};
+
+  constexpr double gridsize = 10;
+  auto nodes = gurka::detail::map_to_coordinates(ascii_map, gridsize);
+  auto admin_map = gurka::buildtiles(nodes, ways, {}, relations, workdir);
+
+  return admin_map;
+}
+
+uint32_t GetContainingState(GraphTileBuilder& tilebuilder,
+                            const std::unordered_multimap<uint32_t, multi_polygon_type>& polys,
+                            const PointLL& ll) {
+  uint32_t index = 0;
+  point_type p(ll.lng(), ll.lat());
+  for (const auto& poly : polys) {
+    if (boost::geometry::covered_by(p, poly.second)) {
+      uint32_t poly_id = poly.first;
+      const Admin& admin = tilebuilder.admins_builder(poly_id);
+      // if state_offset is non-zero, it is a state (not a country)
+      if (admin.state_offset() != 0)
+        return admin.state_offset();
+    }
+  }
+  return 0;
+}
+
+} // anonymous namespace
+
+// Test that BuildAdminFromPBF() works.
+//
+// This test creates a mock pbf AND tile data. Both are required to prove
+// that BuildAdminFromPBF() is working.
+//
+// Given map.pbf, BuildAdminFromPBF() creates test/data/admin.sqlite. Hence,
+// this test needs to prove that the contents of admin.sqlite are correct.
+// Before we get into how I propose to prove correctness, some thoughts must
+// be written.
+//
+// The mock graph we're using has three "relations" (this defines three administrative
+// boundaries). BuildAdminFromPBF() will create three rows in the admin.sqlite::admins
+// table, one for each admin area.
+//
+// GetAdminInfo() will query the tile-data AND the sqlite file to retrieve information
+// about the administrative polygons stored within. You'll see that a map of "polys"
+// is returned - but each "poly" returned is just polygonal data. The remaining admin
+// area metadata is stored in the GraphTileBuilder object that was passed into GetAdminInfo().
+//
+// To prove the admin.sqlite is built correctly, we need to ensure the integral key's
+// stored for each entry in the admins table are correct. We can pass these integral
+// admin keys to GraphTileBuilder::admins_builder(int) to retrieve information about each
+// admin via the returned Admin object.
+//
+// As you can see in the ascii-map, there is a simple two node highway that spans
+// the two states. I thought it'd be interesting to see if I could prove that the
+// node "G" lives in "Colorado" and that "H" lives in "Utah".
+
+TEST(AdminTest, TestAdminPolygonBasic) {
+  // Creates test/data/admin/map.pbf and tile data
+  const std::string workdir = "test/data/admin";
+  auto admin_map = BuildPBFandTiles(workdir);
```

Comment:
if you do this first, before you build admins then the tiles themselves will not have admin information. we need to build the admins first and then build the tiles. can you explain a little more about why we need to do this first? im pretty sure all we need is to build the pbf using the gurka::detail bits. should be approximately like this:

1. make ascii nodes ways relations
2. gurka map_to_coords
3. gurka::detail::build_pbf
4. make config file
5. build admins
6. check that the admin information in the sqlite file looks good
7. build tiles (i get that this still will recreate the pbf, we could either live with that inefficiency or we could add a parameter to specify the existing path to the pbf)
8. check that the admin names on the nodes in the graph are correct

Reply:
Yep, we discussed.  Took your suggestions.